Inviting unregistered users by email

klakhov opened the pull request:
<!-- Raise an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [Contribution guide](https://opencv.github.io/cvat/docs/contributing/). -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable for some reason, then ~~explicitly strikethrough~~ the whole
line. If you don't do that, GitHub will show incorrect progress for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into the [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

azhavoro commented on cvat/apps/organizations/serializers.py:
```diff
@@ -127,3 +146,35 @@ class Meta:
+
+class AcceptInvitationSerializer(RegisterSerializer):
+    def validate_username(self, username):
+        return username
```
Why this metod is needed?

azhavoro commented on cvat/apps/organizations/serializers.py:
```diff
@@ -127,3 +146,35 @@ class Meta:
+        return username
+
+    def validate_email(self, email):
+        return email
```
the same comment as above

klakhov replied:
We need to override `validate_email` because `RegisterSerializer` expects it to be unique, but we already have create user with such email

azhavoro replied:
It's clear

azhavoro commented on cvat/apps/organizations/views.py:
```diff
@@ -205,11 +209,39 @@ def perform_create(self, serializer):
             key=get_random_string(length=64),
-            organization=self.request.iam_context['organization']
+            organization=self.request.iam_context['organization'],
+            request=self.request
```
```suggestion
            request=self.request,
```

azhavoro commented on cvat/apps/organizations/serializers.py:
```diff
@@ -127,3 +146,35 @@ class Meta:
+    def save(self, request, invitation):
+        self.cleaned_data = self.get_cleaned_data()
+        user = invitation.membership.user
+        if "password1" in self.cleaned_data:
```
It seems this condition is always true

klakhov commented on cvat/apps/organizations/models.py:
```diff
@@ -53,18 +58,43 @@ class Meta:
+    def organization_slug(self):
+        return self.membership.organization.slug
+
+    def send(self, request):
```
Currently we get 500 error from server if thats a case. Should we handle it explicitly?

klakhov replied:
Now we will have explicit error:
![image](https://github.com/opencv/cvat/assets/50956430/6798b00f-0e18-48a9-a2c9-52f6b4bb911a)

bsekachev commented on cvat-core/src/api-implementation.ts:
```diff
@@ -106,6 +106,28 @@ export default function implementAPI(cvat) {
         await serverProxy.server.resetPassword(newPassword1, newPassword2, uid, token);
     };
 
+    cvat.server.acceptInvitation.implementation = async (
+        username,
+        firstName,
+        lastName,
+        email,
+        password,
+        userConfirmations,
+        key,
+    ) => {
+        const orgSlug = await serverProxy.server.acceptInvitation(
+            username,
+            firstName,
+            lastName,
+            email,
+            password,
+            userConfirmations,
+            key,
+        );
+
+        return orgSlug;
+    };
```
Not sure that ``server`` is good namespace here.
What do you think about ``cvat.organizations.acceptInvitation``?

bsekachev commented on cvat-core/src/organization.ts:
```diff
@@ -234,24 +319,9 @@ Object.defineProperties(Organization.prototype.members, {
-                    return Promise.resolve();
-                }),
-            );
+            result.results = result.results.map((rawMembership: SerializedMembershipData) => new Membership(
+                rawMembership,
+            ));
```
Instead of redefining ``result.results`` better to create a new identificator. 

const memberships = .... 
memberships.count = result.count;
return memberships;

SpecLad commented on cvat/schema.yml:
```diff
@@ -1531,6 +1531,64 @@ paths:
           description: The invitation has been deleted
+  /api/invitations/{key}/accept:
+    post:
+      operationId: invitations_create_accept
```
Why is this called `invitations_create_accept`? Doesn't seem like it would create an invitation.

klakhov replied:
Fixed

SpecLad commented on cvat/schema.yml:
```diff
@@ -1531,6 +1531,64 @@ paths:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/InvitationWriteRequest'
```
Why does this accept a request body? You shouldn't need to supply anything to accept an invitation (except the identifier of the invitation, which is already passed via the URL path).

klakhov replied:
As we discussed on HLD meeting, we should accept invitation and update user data in one request. Doing this operation in two or three requests can lead to additional problems wich we will have to handle. (invitation expired, deleted, user nickname is already user etc.)

SpecLad commented on cvat/schema.yml:
```diff
@@ -1531,6 +1531,64 @@ paths:
+          content:
+            application/vnd.cvat+json:
+              schema:
+                $ref: '#/components/schemas/InvitationWrite'
```
I'm not sure why this needs to return a response body, but even if it does, shouldn't it be `InvitationRead` rather than `InvitationWrite`?

klakhov replied:
Fixed

SpecLad commented on cvat/schema.yml:
```diff
@@ -1531,6 +1531,64 @@ paths:
+              schema:
+                $ref: '#/components/schemas/InvitationWrite'
+          description: ''
+  /api/invitations/{key}/resend:
```
Same comments here as for `/accept`.

SpecLad commented on cvat/schema.yml:
```diff
@@ -7963,10 +8062,9 @@ components:
-          type: string
-          readOnly: true
-          nullable: true
+          $ref: '#/components/schemas/BasicInvitation'
```
This change looks backwards-incompatible. Is there a way to make it in a backwards-compatible way?

klakhov replied:
There is a way - to request each invitation individually on memberships page, but this leads to doing 10-15 requests, wich is not really good. 
BTW I dont think this field was used in any way, because all the previous invitations were auto accepted.

SpecLad replied:
I was thinking more along the lines of leaving this field as it is, and adding a new field with the `BasicInvitation` type.

> BTW I dont think this field was used in any way, because all the previous invitations were auto accepted.

Even if it wasn't used by the UI, it was still exposed. In particular, the SDK will attempt to parse all fields in a response, even if you don't use some of them, so this change will break it.

klakhov replied:
Ok. I've reverted that change.
Personally, I dont like adding new field like `invitation_data` with invitation details, having a couple of invitation fields looks inconvenient. So my suggestion will be keep it as it was for now.

bsekachev commented:
Is it possible to redirect to login or default tasks page if invitation already accepted? 
Because now I can open this form many times: 
![image](https://github.com/opencv/cvat/assets/40690378/5ab64eb4-9216-4f8d-8184-60617293633c)

After a regular registration user gets authenticated immediately after registration. Here it suggest me to login after. Why is it different? 

Probably it is true, but I am asking just in case. If a user was invited by email, after registration the user's email gets verified without additional actions? 


Maybe we should discuss with somebody, but I believe light blue on orange button is not really contrast:
![image](https://github.com/opencv/cvat/assets/40690378/5a0e910b-312c-4b0d-9466-6a3d9b18263f)
It is unpleasant to look at the button.

bsekachev commented on cvat-ui/src/components/accept-invitation-page/accept-invitation-page.tsx:
```diff
@@ -0,0 +1,50 @@
+        dispatch(acceptInvitationAsync(
+            registerData,
+            invitationParams.key,
+            (orgSlug: string) => {
+                history.replace(`/auth/login?next=/tasks&activateOrganization=${orgSlug}`);
+            },
```
I do not think approach with activate command in query is good enough. Why do we need this? Why can't we activate organization in this callback? Why can't we add org slug to local storage and reload the page? Even not sure that we need to reload, we can look at current org in org watcher where now query parameter is tracked in onmount hook

klakhov replied:
I've checked it, we can set `currentOrganization` and it will be automatically activated without reload, just on login

bsekachev commented on cvat-ui/src/components/register-page/register-form.tsx:
```diff
@@ -34,6 +34,8 @@ export interface RegisterData {
 interface Props {
     fetching: boolean;
     userAgreements: UserAgreement[];
+    predifinedEmail?: string;
```
```suggestion
    predefinedEmail?: string;
```

bsekachev commented on cvat-ui/src/components/register-page/register-form.tsx:
```diff
@@ -34,6 +34,8 @@ export interface RegisterData {
     fetching: boolean;
     userAgreements: UserAgreement[];
+    predifinedEmail?: string;
+    disableNavigation?: boolean;
```
Sounds too abstract
I would propose 

```suggestion
    hideLoginLink?: boolean;
```

or something more specific up to you

klakhov commented on cvat-core/src/user.ts:
```diff
@@ -3,7 +3,7 @@
 // SPDX-License-Identifier: MIT
 
-interface RawUserData {
+export interface RawUserData {
```
Done

klakhov commented on cvat-ui/src/components/organization-page/members-list.tsx:
```diff
@@ -56,6 +60,21 @@ function MembersList(props: Props): JSX.Element {
+                                );
+                                dispatch(
+                                    deleteOrganizationInvitationAsync(organizationInstance, key),
+                                );
```
Yes, its guaranteed. We can delete only org member. Removed second request

klakhov commented:
@bsekachev 
> Is it possible to redirect to login or default tasks page if invitation already accepted? Because now I can open this form many times: 
As we discussed on HLD meeting, We shouldn't do initial request to check invitation status, because after form is filled invitation can become invalid(exired, deleted etc...). But it would be nice feature for the future. 

> After a regular registration user gets authenticated immediately after registration. Here it suggest me to login after. Why is it different?
Maybe we should add it. But currently we consider this form as changing user account info (Like change password, as user account is already created. In password changing logic we redirect user to login form after successful change.) 

> Probably it is true, but I am asking just in case. If a user was invited by email, after registration the user's email gets verified without additional actions?
Yes, thats true

> Maybe we should discuss with somebody, but I believe light blue on orange button is not really contrast: ![image](https://user-images.githubusercontent.com/40690378/272271939-5a0e910b-312c-4b0d-9466-6a3d9b18263f.png) It is unpleasant to look at the button.
Changed it to white:
![image](https://github.com/opencv/cvat/assets/50956430/1fd506d2-5784-4eff-b9f0-d99c45ce8375)

bsekachev commented:
@klakhov 

> As we discussed on HLD meeting, We shouldn't do initial request to check invitation status, because after form is filled invitation can become invalid(exired, deleted etc...). But it would be nice feature for the future.

If discussed on HLD, ok. 
But IMHO user should now if invitation was expired before filling the form, otherwise UX is worse.

klakhov commented on cvat/schema.yml:
```diff
@@ -1531,6 +1531,60 @@ paths:
+      - signatureAuth: []
+      - basicAuth: []
+      responses:
+        '200':
```
Done

klakhov commented on cvat/schema.yml:
```diff
@@ -6100,6 +6154,35 @@ components:
       - description
       - name
       - version
+    AcceptInvitationRequest:
```
There was a mistake, this serializer doesnt accept `email` field, fixed that. Also, it has different save logic (it gets user email from the invitation, doesn't create user as its already created), so I dont think we can use `RegisterSerializerEx` here

klakhov commented on cvat/schema.yml:
```diff
@@ -1531,6 +1531,60 @@ paths:
+          content:
+            application/vnd.cvat+json:
+              schema:
+                type: string
```
Done

nmanovic commented:
Please let me play a littile bit with the PR. Probably, I have some questions.
